### PR TITLE
proposal for P4RT-2.1 Mastership test

### PR DESCRIPTION
**Pull request by Hobbydos**

### CODE REVIEW

**Comment by Hobbydos** [comment]
cc @xw-g @robshakir

File: feature/experimental/p4rt/tests/p4rt_mastership/p4rt_mastership.MD
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_mastership/p4rt_mastership.MD:9]: Primary/Secondary over master/slave?

**Review by Ankur19 — CHANGES_REQUESTED** [review]
Looks Good except a few NITs

File: feature/experimental/p4rt/tests/p4rt_mastership/p4rt_mastership.MD
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_mastership/p4rt_mastership.MD:9]: @Ankur19, Should we rename the test to P4RT election instead of mastership to keep the trend of inclusive language?

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:564]: Single testArg. Doesn't need a table driven test and the loop over testCases.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:488]: The only variable I see are the "desc" and the "lowID". Can you just use the table for those and create a single testArg

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:433]: Can we only have the values that are changing in the table here and then add the constant items later in `for test:= range testCases{}` loop

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:382]: Similar to the other comments below. Can we only add the variables in the table here.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:326]: Same here

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:288]: No need of table here.

**Review by Ankur19 — CHANGES_REQUESTED** [review]

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:564]: I can do that, however if there are other tests to be added in the future we'll have to re-implement table driven test  + loop.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (robshakir) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:65]: Generally we'd call these `inXXX` for input and `wantXXX` for expected values.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (robshakir) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:74]: Suggest to just define `component` here as a pointer since you never use it in a manner that requires it not to be a pointer.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (robshakir) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:96]: Can we define here what the return argument is? i.e., the arbitration status code.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (robshakir) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:293]: (nit) remove blank line,

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (robshakir) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:302]: Do we expect to have >1 tests within this test case going forward? I guess the `t.Run` here is intended to just allow us to get a better failure report (since it seems like you couldn't just run this test case without the setup).

**Comment by robshakir** [comment]
Thanks for this contribution -- generally LGTM, just some cleanup around the code that I think we need and to check some of the definitions.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:96]: @robshakir , did you mean the return on line 93 or the comment from the function starting on line 96?

File: feature/experimental/p4rt/tests/p4rt_election/README.md
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/README.md]: Pushed a commit with changes requested.

**Comment by Hobbydos** [comment]
Went through and added suggested changes.

**Review by robshakir — COMMENTED** [review]
This LGTM -- with one small pending question and one suggestion.

@xw-g -- do you want to take a look before we go ahead and merge?

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go]: Made the suggested changes.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:256]: Logic error here. Should it not be args.wantWrite==true?

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:313]: Is this a requirement? Couldn't find anything related to this here: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-updates

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:256]: We're only hitting this block when the write actually fails (err != nil) and if args.wantWrite is false, then it's an expected error, hence the t.Logf

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:313]: There was some debate on this, however the lack of clarification of the spec is what prompted for a decision. @robshakir, thoughts?  I checked on a few vendors and they considered (0,0) as invalid election ID as well.

The closest thing I found in the spec under `16.2. Client Arbitration Update` was: 
> The sender need not specify an election_id. If the election_id is not specified, the sender's election_id is considered lower than any election_id, and the sender will thus never become primary. This way, a controller can choose to be a backup controller, in order to avoid “flapping” (if a backup controller connects to the switch shortly before the actual primary controller, therefore becoming primary temporarily).

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: Should it be 3?

```
If the election_id is set and is already used by another controller (excluding the controller making the request) for the same (device_id, role), the P4Runtime server shall terminate the stream by returning an INVALID_ARGUMENT error.
```

error code for INVALID_ARGUMENTis 3

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (rsindhujha) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:313]: In master Arbitration, election ID zero need no be treated as a special case. The P4RT server should be able to accept the connection and client with election ID 0 may become primary, if its the only client getting connected.

**Comment by rsindhujha** [comment]
In the below testcase(func TestPrimaryReconnect(t *testing.T)), the client is getting disconnected after every iteration. This means the connection should be successful for each iteration with election ID 100, 90 and 100 again. Since there is only one client exists within a given iteration, for iteration 2, the wantStatus should be 0 instead of 5?

https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L327

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (rsindhujha) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: For the below Testcase, sometimes we are observing that the testcase did not receive the status code 3, as the P4RT server already disconnects the streamChannel after sending the MasterArbitration Response. 

 func TestDuplicateElectionID(t *testing.T) { 

Reference from the specs (5.3.2.4)
If the election_id is set and is already used by another controller (excluding the controller making the request) for the same (device_id, role), the P4Runtime server shall terminate the stream by returning an INVALID_ARGUMENT error.

**Comment by rsindhujha** [comment]
For the last testcase, the script is sending 2nd Arbitration Request with lower electionID over the same StreamChannel. This Arbitration Request should be accepted by the server, since there is no other client. In this case the expected Status code should be 0 and wantWrite True

https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L567

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: > In the below testcase(func TestPrimaryReconnect(t *testing.T)), the client is getting disconnected after every iteration. This means the connection should be successful for each iteration with election ID 100, 90 and 100 again. Since there is only one client exists within a given iteration, for iteration 2, the wantStatus should be 0 instead of 5?
> 
> https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L327

I think there might be a disconnect here, the server would keep track of the Primary Election ID, should a connection come in with anything < Primary Election ID - Even if the Primary is disconnected, the error MSG should be 5.

From Section 16.2 - https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
> the P4Runtime server keeps track of the highest election_id that it has ever received. If a controller's election_id is equal to the highest election_id that the server has ever received, that controller is the primary. All other controllers are backups. Note that it is possible that all controllers are backups, and that there is no primary. There can be at most one primary, because for any given (device_id, role), each connected controller has a unique election_id.

And Section 5.4: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
>If there is a primary:
For the primary, status is OK (with status.code set to google.rpc.OK).
For all backup controllers, status is set to non-OK (with status.code set to google.rpc.ALREADY_EXISTS).
Otherwise, if there is no primary currently, for all backup controllers, status is set to non-OK (with status.code set to google.rpc.NOT_FOUND).

GRPC status codes for ref: https://developers.google.com/maps-booking/reference/grpc-api/status_codes

**Comment by Hobbydos** [comment]
> In the below testcase(func TestPrimaryReconnect(t *testing.T)), the client is getting disconnected after every iteration. This means the connection should be successful for each iteration with election ID 100, 90 and 100 again. Since there is only one client exists within a given iteration, for iteration 2, the wantStatus should be 0 instead of 5?
> 
> https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L327

> In the below testcase(func TestPrimaryReconnect(t *testing.T)), the client is getting disconnected after every iteration. This means the connection should be successful for each iteration with election ID 100, 90 and 100 again. Since there is only one client exists within a given iteration, for iteration 2, the wantStatus should be 0 instead of 5?
> 
> https://github.com/openconfig/featureprofiles/blob/b389c9712e2b2a88758b11dd4cc1dac12bf0ef4c/feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go#L327

I think there might be a disconnect here, the server would keep track of the Primary Election ID, should a connection come in with anything < Primary Election ID - Even if the Primary is disconnected, the error MSG should be 5.

From Section 16.2 - https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
> the P4Runtime server keeps track of the highest election_id that it has ever received. If a controller's election_id is equal to the highest election_id that the server has ever received, that controller is the primary. All other controllers are backups. Note that it is possible that all controllers are backups, and that there is no primary. There can be at most one primary, because for any given (device_id, role), each connected controller has a unique election_id.

And Section 5.4: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-arbitration-notification
>If there is a primary:
For the primary, status is OK (with status.code set to google.rpc.OK).
For all backup controllers, status is set to non-OK (with status.code set to google.rpc.ALREADY_EXISTS).
Otherwise, if there is no primary currently, for all backup controllers, status is set to non-OK (with status.code set to google.rpc.NOT_FOUND).

GRPC status codes for ref: https://developers.google.com/maps-booking/reference/grpc-api/status_codes

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: I'm currently looking into this case, as you mention the streamChannel gets disconnected right after the MasterArbitration Response, making it challenging to capture the Arbitration response status code. Might have to look at changing the client, but it's a WIP.

Comment (rsindhujha) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: Thanks @Hobbydos

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:313]: @rsindhujha, IIRC there was a conversation with @robshakir where we had concluded that zero election ID was not to become primary even if it was the only client getting connected.  Perhaps Rob can share thoughts here.

**Comment by xw-g** [comment]
> This LGTM -- with one small pending question and one suggestion.
> 
> @xw-g -- do you want to take a look before we go ahead and merge?

no, I'll let you drive this one. we already have 3 reviewers on this one. That's good enough :)

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:129]: You can now get the StreamTermErr after sendinding Msg.

```
if _, _, arbErr := client.StreamChannelGetArbitrationResp(&streamName, 1); arbErr != nil {
  if err := p4rtutils.StreamTermErr(client.StreamTermErr); err != nil {
	  return err
  }
  return fmt.Errorf("errors seen in ClientArbitration response: %v", arbErr)
}
```

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: @Hobbydos please refer to other tests to see how to capture Stream Termination Errors and then update to the correct status codes to validate

Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:454]: Thanks @Ankur19  - I plan to revisit this shortly. It seems like I'll have to change how the p4rt node name is passed, too.
Ref:

Tests should get the P4RT Node Name by walking the Components OC tree. Components of type INTEGRATED_CIRCUIT should have child Components of type PORT. These PORT Components can be mapped to currently reserved Interfaces using the hardware-port leaf in the Interfaces tree. Such an implementation already exists in p4rtutils library: p4rtutils.P4RTNodesByPort().

If P4RT Node Names cannot be resolved by walking the Components tree, use deviation flag --deviation_explicit_p4rt_node_component and pass the node names through args --arg_p4rt_node_name_1, --arg_p4rt_node_name_2.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:129]: I tried to look at `StreamTermErr` but it seemed like it was `nil`. Is it possible that the stream gets destroyed before reaching the code? 
Here's a view from DLV in VSCode of the args
image
 
Code Snippet (don't mind the `return 3`, I was putting it there for debug purposes.
```go
func getRespCode(args *testArgs) (int32, error) {
	handle := args.handle
	if handle != nil {
		// Grab Arb Response to look at status code
		_, arbResp, arbErr := handle.StreamChannelGetArbitrationResp(&streamName, 1)
		if arbErr != nil {
			if err := p4rtutils.StreamTermErr(args.handle.StreamTermErr); err != nil {
				return 3, err
			}
			return 0, fmt.Errorf("Errors seen in ClientArbitration response: %v", arbErr)
		}
		// Handle exception when GetCode is empty.
		respCode := arbResp.Arb.GetStatus().GetCode()
		return respCode, nil
	}
	return 0, errors.New("Missing Client")
}
```

Comment (Ankur19) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:129]: I think it is because of how you initialize `P4RTClient`.

* If you do `p4rt_client.P4RTClient{}` then `StreamTermErr` channel is not initialized.

The proper way would be using the method `p4rt_client.NewP4RTClient()`.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:97]: Change to `p4rt_client.NewP4RTClient()` method. This would initialize `StreamTermErr`

Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:97]: Thanks @Ankur19, changes worked. Just working on a few more things and should be in a decent place to push something for this week.

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go]: @Ankur19 The requested changes have been added

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:151]: Please use status and code package here. 

https://pkg.go.dev/google.golang.org/grpc/status

**Review by Ankur19 — CHANGES_REQUESTED** [review]

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:151]: sounds good, will add

Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:151]: Just to make sure I'm not going off the deep-end and adhering to the standard, you mean something like this, correct?

```go
				if status.Code(err) == codes.InvalidArgument {
					return int32(codes.InvalidArgument), err
				}
```

Comment (Ankur19) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:151]: No need to validate `InvalidArgument` here. Just return whatever error was returned by the server. 

```
If err != nil {
   return int32(status.Code(err))
}
```

Comment (Hobbydos) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:151]: SG - Thanks for clarification

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [inline feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:150]: err would always be nil here. No need to check again. 

```
if err := p4rtutils.StreamTermErr(args.handle.StreamTermErr); err != nil {
				return int32(status.Code(err)), err
}
```

File: feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go
Comment (Ankur19) [reply feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go:325]: This looks somewhat relevant: https://p4.org/p4-spec/p4runtime/main/P4Runtime-Spec.html#sec-zero-as-reserved-value

**Review by Ankur19 — DISMISSED** [review]
LGTM!

**Comment by xw-g** [comment]
@liulk is the rundata checking failure a blocker to this merge? Thanks.

**Comment by xw-g** [comment]
@Hobbydos can you help run `go run ./tools/addrundata --fix` (asked @liulk offline). Thanks!

**Comment by Hobbydos** [comment]
> @Hobbydos can you help run `go run ./tools/addrundata --fix` (asked @liulk offline). Thanks!

@xw-g Ran and committed - cc @liulk

**Comment by xw-g** [comment]
Since it's a new test, I"m skipping the internal presumit run. Will merge upon the github check passes.

**Comment by xw-g** [comment]
@Ankur19 and @robshakir gave me LGTM offline.

**Review by xw-g — APPROVED** [review]

**Pull request merged**
